services/github/src/handlers/status.js

```diff
@@ -1,30 +1,201 @@
+    this.name = name;
+    this.head_sha = head_sha;
+    this.external_id = external_id;
+    this.details_url = details_url;
+
+    this.status = status;
+    this.conclusion = conclusion;
+
+    this.output = new GithubCheckOutput({
+      title: output_title,
+      summary: output_summary,
+      text: output_text,
+      annotations: output_annotations,
+    });
+  }
+
+  /**
+   * Github check run conclusion and completed_at fields should only be sent
+   * when task was completed, otherwise github will
+   * automatically resolve check run as completed
+   */
+  getStatusPayload() {
+    const { status, conclusion } = this;
+    const resolution = { status };
+    if (status === CHECK_RUN_STATES.COMPLETED) {
+      resolution.conclusion = conclusion;
+      resolution.completed_at = new Date().toISOString();
+    }
+    return resolution;
+  }
+
+  getCreatePayload() {
+    const { owner, repo, name, head_sha, external_id, details_url, output } = this;
+
+    return {
+      owner,
+      repo,
+      name,
+      head_sha,
+      external_id,
+      details_url,
+      ...this.getStatusPayload(),
+      output: output.getPayload(),
+    };
+  }
+
+  getUpdatePayload() {
+    const { owner, repo, check_run_id, output } = this;
+
+    if (!this.check_run_id) {
+      throw new Error('Updating check run without check_run_id');
+    }
+
+    return {
+      check_run_id,
+      owner,
+      repo,
+      ...this.getStatusPayload(),
+      output: output.getPayload(),
+    };
+  }
+
+  getRerequestPayload() {
+    const { owner, repo, check_run_id } = this;
+
+    if (!this.check_run_id) {
+      throw new Error('Rerequesting check run without check_run_id');
+    }
+
+    return {
+      check_run_id,
+      owner,
+      repo,
+    };
+  }
+}
+
+/**
+ * Post updates to GitHub, when task is being created or its status changes.
+ * Uses Checks API
+ *
+ * @param message exchange message
+ *  One of:
+ *  taskDefined: https://docs.taskcluster.net/docs/reference/platform/taskcluster-queue/references/events#taskDefined
+ *  taskRunning: https://docs.taskcluster.net/docs/reference/platform/taskcluster-queue/references/events#taskRunning
+ *  taskCompleted: https://docs.taskcluster.net/docs/reference/platform/taskcluster-queue/references/events#taskCompleted
+ *  taskFailed: https://docs.taskcluster.net/docs/reference/platform/taskcluster-queue/references/events#taskFailed
+ *  taskException: https://docs.taskcluster.net/docs/reference/platform/taskcluster-queue/references/events#taskException
+ * @returns {Promise<void>}
  **/
 async function statusHandler(message) {
-  let { taskGroupId, state, runs, taskId } = message.payload.status;
-  let { runId } = message.payload;
-  let { reasonResolved } = runs[runId];
+  const { taskGroupId, state, runs, taskId } = message.payload.status;
+  const { runId } = message.payload;
+  const { reasonResolved } = runs[runId] || {};
+  const taskDefined = state === undefined;
+
+  await qLock.acquire(taskId);
```

Comment:
this ensures we only run single status update for a given task. Consequent handlers will pause here until the previous is done.